Added tinyurl function with config options

brasstax opened the pull request:
Hi there,

Another channel I was a part of had a bot that would shorten long URLs using TinyURL. I think this is a great feature because I use irssi over SSH so I can't click on links, and copying long links has a tendency of not working out for me.  I've replicated that feature in urls.py as an optional feature; if the user includes a `shorten_url_length` attribute with a value of greater than 0 in the URL section of the config, the bot will call TinyURL's API to shorten a link longer than `shorten_url_length` and then include that result in `title_auto`.

As an example:

```
21:36 <@Casull> https://github.com/sopel-irc/sopel/wiki/Module-Configuration
21:36 < lp0-dev> [ Module Configuration · sopel-irc/sopel Wiki · GitHub ] - github.com 
          ( http://tinyurl.com/ybusdxop )

```

brasstax commented:
>I'd really like to ask that tests be added, but that's not really ideal from TinyURL's side, as they generate a new short URL keyword every time the same long URL is submitted. Every CI job (and local test run) would create a new TinyURL pointing to the test address.

I did some experimentation and it seems the TinyURL API uses the same short URL keyword every time. As a test, would you mind running `curl https://tinyurl.com/api-create.php?url=https://www.google.com` and seeing if you end up with `http://tinyurl.com/389lo` as the result?  I've tested it from my desktop and the host where I'm running a production bot and it seems to return the same result.

Since this is the case, I can make this testable. Do you have a suggestion for testing with functions that aren't explicitly IRC commands, like `.title`?  The `@example` decorator looks like it assumes that the command being tested has a way for the user to call it, but `get_tinyurl` isn't really meant to be user-facing.  I glanced at the unit-testing section in the wiki but I didn't see anything, either. While I _could_ make this a user-facing command I'm not sure if there's a point to doing that.

>Along with caching seen URLs in bot.memory to avoid duplicate API calls
Also a good idea. I'll add it to my todo list.

brasstax commented on sopel/modules/url.py:
```diff
@@ -46,6 +48,11 @@ def configure(config):
+        'shorten_url_length',
+        'Enter how many characters a URL should be before the bot puts a'
+        ' shorter version of the URL in the title as a TinyURL link'
+    )
```
Done.

brasstax commented on sopel/modules/url.py:
```diff
@@ -139,11 +151,12 @@ def process_urls(bot, trigger, urls):
-    Return a list of (title, hostname) tuples for each URL which is not handled by
-    another module.
+    Return a list of (title, hostname) tuples for each URL which is not handled
+    by another module.
```
Also done.

brasstax commented on sopel/modules/url.py:
```diff
@@ -231,6 +249,13 @@ def get_hostname(url):
 
+def get_tinyurl(url):
+    """ Returns a shortened tinyURL link of the URL. """
+    tinyurl = "http://tinyurl.com/api-create.php?url=%s" % url
```
No particular reason; I've swapped it to HTTPS because it's a good idea. I've also updated the script to return HTTPS links since it turns out TinyURL's API returns HTTP links regardless if you call it using HTTP or HTTPS.

dgw replied:
That's very silly of TinyURL, but I appreciate that you took care of it!

dgw commented:
Interestingly, when I tested shortening a test URL (not google.com) the other day, TinyURL gave me different keywords each time. Though I was using a browser, not `curl`, so maybe that was my mistake. Mostly, though, I was looking at `is.gd` instead because it's shorter than `tinyurl.com`.

The way to test this is probably attaching an example to the function that's actually called when the bot sees a URL, and do it that way, assuming I'm correct that `@example` will work that way. Ultimately we want to test the whole URL-handling code path, including the title fetching and shortener, etc. Starting from the URL being passed to the bot as an input line is the best way to do that.

I'm not sure of the best way to activate the shortener function from the test, though, since the default setting is off. I think it's fine to leave tests to another PR, since it's not removing an existing test and there's lots of stuff that needs tests added besides this. (I suspect the real best way forward will be adding some way of defining a "test default" for settings if it needs to be different from the user-facing default, and/or a runtime condition modules can check that's more abstract than `if 'pytest' in sys.modules`.)

Since this PR now will need to be interactively rebased before merging, are you set up to do that? It doesn't look like you're committing from GitHub's web flow, so you're probably using git in a terminal, but if you need instructions on how to do it I'm happy to write a guide once this is ready to go.

dgw commented on sopel/modules/url.py:
```diff
@@ -231,6 +259,17 @@ def get_hostname(url):
+    # result an HTTPS link (the third argument is so that it only
+    # replaces the first HTTP portion of the TinyURL and not, say,
+    # a hypothetical short URL with HTTP in it.
+    return res.text.replace("http", "https")
```
You referenced a "third argument" in the comment block, but passed only two arguments to the function. That said, you can make this even safer by replacing `http://` with `https://` instead.

brasstax replied:
Yeah, you're right. This is done.

brasstax commented:
Whoops, I added some more stuff before I saw your last comment. Could you take a look at what I've done?

I incorporated using Sopel's memory to store shortened URLs, but I wanted to make it neat, so I had Sopel create a new `bot.memory['shortened_urls']` entry if the `shortened_urls` entry didn't exist. However, I also wanted to be thread-safe, and I didn't want to litter `bot.memory` with a bunch of entries for `%s_shortened_url`s, so I created a separate `tools.SopelMemory()` dict to hold shortened_urls. Is that the right way to go about it?

dgw commented:
Your `bot.memory['shortened_urls']` usage looks good, it matches what several other modules do to initialize their memory space.

This looks ready to squash. I'll give it a final look when it's time to merge. Haven't decided if this is going into 6.6.0 yet (but it probably will).

brasstax commented:
>Interestingly, when I tested shortening a test URL (not google.com) the other day, TinyURL gave me different keywords each time. Though I was using a browser, not curl, so maybe that was my mistake. 

That's very interesting to hear. I tried using the browser now as well as `curl` for another long URL (My example URL used is https://en.support.wordpress.com/markdown-quick-reference/) and I'm getting back the same result every time (https://tinyurl.com/qafujs9). Running it in the API also seems to give me the same URL as well (https://tinyurl.com/api-create.php?url=https://en.support.wordpress.com/markdown-quick-reference/). Would you mind opening the tinyURL's api-create link in a browser and seeing if you end up with `https://tinyurl.com/qafujs9` as the result?

Thank you for your patience in putting up with my slight paranoia here.

>Since this PR now will need to be interactively rebased before merging, are you set up to do that? It doesn't look like you're committing from GitHub's web flow, so you're probably using git in a terminal, but if you need instructions on how to do it I'm happy to write a guide once this is ready to go.

Guilty as charged.  I could probably Google it (and I am as we speak) but if you have a preferred way of going about it I'm all ears.

dgw commented:
> Would you mind opening the tinyURL's api-create link in a browser and seeing if you end up with `https://tinyurl.com/qafujs9` as the result?

I did. Now I have no idea what was up the other day, maybe I mistyped it once and omitted a trailing slash? Seems to be perfectly stable now, anyway.

> Guilty as charged. I could probably Google it (and I am as we speak) but if you have a preferred way of going about it, I'm all ears; it's my first time actually rebasing my projects and I'm always down to learn how to do my git flow better.

It's slightly more complicated because you opened this PR from your `master` branch, but still easy enough. You'll want to have a rebase target, so first add this repo as the remote 'upstream', then fetch its branches. Finally, do the actual rebase:

* `git remote add upstream git@github.com:sopel-irc/sopel.git`
* `git fetch upstream`
* `git rebase -i upstream/master`

That last command will open a text editor (likely the same one you get when you run `git commit`) and list the commits on your current branch. The default action is just `pick`, which applies the commit without any changes. You'll want to change the first one to `reword` (more on that in a sec), and the rest to `squash`. Save the file and close your editor to start the rebase.

With `reword`, Git will stop to let you rewrite that commit's log message. I'm asking you to do that so you can match Sopel's commit message guidelines: "Our convention is to place the name of the thing you're changing in at the beginning of the message, followed by a colon" (`CONTRIBUTING.md`). In this case, that means something like "url: add optional shortening of long URLs" as the first line. In the body of the commit message you can elaborate, or wait until the next step.

When you save (and close) the new commit message for the first commit, Git will then present you with another commit message to edit. This one will have comments to tell you what's what, but in short it's a combination of all the commit messages from the commits you're squashing. You won't need most of them, but I suggested using `squash` instead of `fixup` because it shows you all this extra detail that you can remove if it's not needed. Usually this is where I write the message I want to keep in the log at the top of the file, using details from the later ones, then delete everything below the first comment before saving. You might find another workflow more appealing; that's just how I usually do this kind of thing.

Once you finish writing the "real" commit message (and deleting stuff that doesn't need to be in the log, too), save it and close the editor. Git will finish the rebase and update your branch (`master`). Here I usually verify what the branch looks like with `git log`. If it looks good, `git push -f` will overwrite the branch on GitHub with what you just made.

----

That turned out longer than I expected, but I hope it's not so long that it's confusing. And don't worry if you mess something up somewhere and can't undo the mistake: I have a backup copy of your branch just in case. :)

brasstax commented:
Rebase done.

dgw commented:
I was about to merge this, and then realized that if `get_tinyurl()` throws an exception—if, for example, TinyURL goes down—it would make the bot spit out an error and abort expanding the title.

@brasstax How annoyed would you be if I asked you to rework the code so it catches such cases and proceeds with the title expansion anyway, sans tinyurl? I can certainly do it, but not necessarily before the next release of Sopel.

dgw commented on sopel/modules/url.py:
```diff
@@ -231,6 +259,18 @@ def get_hostname(url):
+    tinyurl = "https://tinyurl.com/api-create.php?url=%s" % url
+    try:
+        res = requests.get(tinyurl)
+    except requests.exceptions.ConnectionError:
```
The `ConnectionError` exception won't cover all cases. Timeouts, misconfigurations that cause excessive redirects, invalid/malformed HTTP responses… these all have different exceptions in Requests.

But! We can catch them all by catching the base class:

```suggestion
    except requests.exceptions.RequestException:
```

dgw commented on sopel/modules/url.py:
```diff
@@ -231,6 +259,18 @@ def get_hostname(url):
+    """ Returns a shortened tinyURL link of the URL. """
+    tinyurl = "https://tinyurl.com/api-create.php?url=%s" % url
+    try:
+        res = requests.get(tinyurl)
```
For additional robustness, raise an exception for HTTP error status codes as well:

```suggestion
        res = requests.get(tinyurl)
        res.raise_for_status()
```

This will guard against changes in the TinyURL API that result in HTTP 4xx errors, or API downtime that is severe enough to return HTTP 5xx errors but not so severe that the API cannot accept and process connections.

dgw commented:
> (The test failed but it looks like that's from #1276 )

Indeed. Why that commit wound up in this PR, though… It shouldn't be here. :joy_cat:

dgw commented:
> If it's all right with you, I'd like to close this request and open a new request where I _properly_ make a pull request from my `tinyurls` branch with the tinyURL feature.

I'm not _against_ doing that, but this PR will be ready to go if you apply the two suggestions I put in the latest code review and squash it down (removing the unrelated changes). Then you can have your `master` branch back to normal _and_ have one less open PR to worry about. :smile_cat:

If you want to preserve your current `master`, you can `git push origin master:backup-master`. Then clean up `master`. I'll merge the PR within hours, and then you can restore from the backup branch.

Opening a new PR is a fine course of action, but this one is so close (and there's so much context in this PR from code reviews and comments) that I think it's worth just finishing out this PR instead. Whaddaya say?

dgw commented:
Well… I took too long writing my response. The new PR exists already.
